elastic/kibana
# [Upgrade Assistant] New ES deprecations page

This PR updates the ES deprecations page in Upgrade Assistant to align with the new design prototype.

It supports 4 different deprecation "types":
1. Reindex old indices
2. Upgrade or delete old ML job model snapshots
3. Remove deprecated index settings
4. Default - any other deprecation that doesn't have an automatic resolution associated with it.

This is representative in the code under `es_deprecations/deprecation_types`.

**Technical implementation details:**
With the new design, we need to be able to track the status of ML and reindex actions at the row level. This logic was previously encapsulated within a button component ([example](https://github.com/elastic/kibana/blob/master/x-pack/plugins/upgrade_assistant/public/application/components/es_deprecations/deprecations/reindex/button.tsx)). This is no longer possible since we need to open the flyout via the "Issue" column and also provide the status of the resolution via the "Resolution" column. In order to achieve this, I built a custom table using EUI components rather than `EuiMemoryTable` that wraps each applicable row with a context provider. Long term, I think it's worth evaluating if we can keep track of all ML and reindex statuses in a single context, rather than per index name and per ML job ID. However, this requires a fairly significant refactoring that I think is out of scope for this PR.

**How to test**
Unfortunately, this can only be tested on `master` with mocked data. That said, I have created a parallel branch based on `7.x` that can be used to test the changes against real deprecations.

1. Check out [my 7.x branch](https://github.com/alisonelizabeth/kibana/tree/7/ua%2Fes_deprecation_design_updates) and bootstrap.
2. Unzip [6.8.16-data.zip](https://github.com/elastic/kibana/files/7007813/6.8.16-data.zip) and start up ES as follows: `yarn es snapshot -E path.data=/path/to/6.8.16-data`. This contains indices and ML job model snapshots created on 6.8. This will trigger the reindex and ML deprecations.
3. Start up Kibana, `yarn start`.
4. Navigate to Console to trigger other 7.x deprecations. Run the following commands:
```
// Triggers the deprecated index settings flyout
PUT deprecated_settings
{
  "settings": {
    "translog.retention.size": "1b",
    "translog.retention.age": "5m",
    "index.soft_deletes.enabled": true
  }
}

```

```
// Deprecation with no automatic resolution
PUT /nested_multi_fields
{
   "mappings":{
      "properties":{
         "text":{
            "type":"text",
            "fields":{
               "english":{
                  "type":"text",
                  "analyzer":"english",
                  "fields":{
                     "english":{
                        "type":"text",
                        "analyzer":"english"
                     }
                  }
               }
            }
         }
      }
   }
}

```

```
// Deprecation with no automatic resolution
PUT field_names_enabled
{
  "mappings": {
    "_field_names": {
      "enabled": false
    }
  }
}

```
5. Finally, you're ready to test! Navigate to Upgrade Assistant --> Elasticsearch deprecations.
    - Test search, pagination, filtering
    - Go through the different deprecations and verify you are able to resolve the ones which provide automatic resolution.
    - After resolving an issue, verify the status has been updated. Reopen the flyout, you should not be able to "re-resolve" the deprecation (read-only). Refreshing the data should remove the deprecation from the list.

**Deferred work**
The following work will be completed in separate PRs:

- Functional tests
- a11y tests
- UI support for `resolve_during_rolling_upgrade` field - waiting on more information from the ES team about this
- "Backup your data" button - waiting on additional feedback as to whether it's necessary to show on this page or not
- Bulk reindexing callout - waiting on copy suggestions
- Reindexing flyout - the flyout was largely unchanged in this PR. Any design updates will be treated as a separate piece of work.

**Screenshots**

<details>
<summary>Overview of changes</summary>
<img width="1781" alt="Screen Shot 2021-08-17 at 12 32 53 PM" src="https://user-images.githubusercontent.com/5226211/129774874-30afb650-dd48-44f9-8ddd-1615c4d62d9f.png">

// Reindex flyout
<img width="1368" alt="Screen Shot 2021-08-17 at 12 33 38 PM" src="https://user-images.githubusercontent.com/5226211/129774917-13cd35c9-8ea9-4b07-a321-5f4ecb82662f.png">

// Deprecated index settings flyout
<img width="1349" alt="Screen Shot 2021-08-17 at 12 35 18 PM" src="https://user-images.githubusercontent.com/5226211/129774965-c4092d0e-a798-4e91-993f-2443659b722d.png">

// Default flyout (deprecation does not have automatic resolution)
<img width="1372" alt="Screen Shot 2021-08-17 at 12 34 25 PM" src="https://user-images.githubusercontent.com/5226211/129775039-b2bdac9f-dd24-4299-bf7b-d8a22b5827f0.png">

// ML flyout
<img width="1495" alt="Screen Shot 2021-08-17 at 1 43 46 PM" src="https://user-images.githubusercontent.com/5226211/129775147-ec329a06-49d5-4e34-84f2-9c7c56c5da8e.png">

// Some resolution states
<img width="1599" alt="Screen Shot 2021-08-17 at 12 38 42 PM" src="https://user-images.githubusercontent.com/5226211/129775276-65626d59-0db4-4826-a63e-7e3134cd5e95.png">

<img width="1462" alt="Screen Shot 2021-08-17 at 12 38 56 PM" src="https://user-images.githubusercontent.com/5226211/129775283-3caa7b33-a79e-4da8-ad62-8be7d4940da4.png">

<img width="1591" alt="Screen Shot 2021-08-17 at 12 37 52 PM" src="https://user-images.githubusercontent.com/5226211/129775459-0eba9727-31a3-4219-abe6-ee74775b7c52.png">

// Tooltip on Resolution type
<img width="382" alt="Screen Shot 2021-08-17 at 12 39 22 PM" src="https://user-images.githubusercontent.com/5226211/129775342-eabce699-ca6d-4a51-ae6c-8ff8307aca34.png">

// No deprecations
<img width="720" alt="Screen Shot 2021-08-17 at 1 49 05 PM" src="https://user-images.githubusercontent.com/5226211/129775857-bc8f38a8-3788-4ad6-aeca-5f26e0bab94b.png">

// Invalid search
<img width="1617" alt="Screen Shot 2021-08-17 at 1 48 00 PM" src="https://user-images.githubusercontent.com/5226211/129775886-7e8ed07b-e369-4b91-99bc-c437c05ff5d4.png">

// Initial loading state
<img width="421" alt="Screen Shot 2021-08-17 at 1 48 45 PM" src="https://user-images.githubusercontent.com/5226211/129775920-84de1fb1-2ee6-4afd-851c-4dcbf5519337.png">

// No search results
<img width="1629" alt="Screen Shot 2021-08-17 at 1 48 08 PM" src="https://user-images.githubusercontent.com/5226211/129775942-d9f993f4-08d9-4b3c-b402-9169c09e0aa5.png">

</details>
